server/container/util/src/test/java/org/apache/james/util/ReactorUtilsTest.java

```diff
@@ -116,9 +110,9 @@ void throttleDownStreamConcurrencyShouldNotExceedWindowMaxSize() {
                     .flatMap(i -> Mono.delay(windowDuration.multipliedBy(2)).thenReturn(i))
                     .flatMap(i -> Mono.fromRunnable(ongoingProcessing::decrementAndGet).thenReturn(i));
 
-            ImmutableList<Integer> ongoingProcessingUponComputationStart = Throttler.forOperation(longRunningOperation)
-                .window(windowMaxSize, windowDuration)
-                .throttle(originalFlux)
+            ImmutableList<Integer> ongoingProcessingUponComputationStart = originalFlux
+                .transform(ReactorUtils.<Integer>throttle().elements(windowMaxSize).per(windowDuration))
+                .flatMap(longRunningOperation)
```

Comment:
Can you explain to me why `throttleDownStreamConcurrencyShouldNotExceedWindowMaxSize` still works?

Reply:
it probably doesn't but maybe you have an idea of why if you put this comment? I couldn't find why it broke

Reply:
The way I understand it is that `transform` don't limit downstream concurrency, hence I was really curious to know if that test pass.

Reply:
![Screenshot from 2020-06-08 10-19-08](https://user-images.githubusercontent.com/708318/84008039-c37e2480-a960-11ea-84a5-ec3a2593b06e.png)
 from https://projectreactor.io/docs/core/release/reference/

How did you saw it's doing anything differently than inline operators?

Reply:
Your operation don't wrap the downstream flatMap request were we do actually perform the operation we want to throttle. If the prossessing is slower than the rate of the throttling you would end up doing more concurrent operations than what you would like, unless you apply some form of limitations to it.

That's what the test enforces.

Also, from what I understand the downstream flatMap needs to be part of the 'extracted operator chain'.

Reply:
A syntax addressing this might be (suggestion)

```
originalFlux
    .transform(ReactorUtils<Integer>throttle()
        .elements(windowMaxSize)
        .per(windowDuration)
        .forOperation(longRunningOperation))
```